src/insertItem.ts

```diff
@@ -0,0 +1,449 @@
+        if (!template) {
+            return;
+        }
+        ext.outputChannel.appendLine("Insert item");
+        switch (sectionType) {
+            case TemplateSectionType.Functions:
+                if (await this.insertFunction(template, textEditor)) {
+                    vscode.window.showInformationMessage("Please type the output of the function.");
+                }
+                break;
+            case TemplateSectionType.Outputs:
+                if (await this.insertOutput(template, textEditor)) {
+                    vscode.window.showInformationMessage("Please type the the value of the output.");
+                }
+                break;
+            case TemplateSectionType.Parameters:
+                if (await this.insertParameter(template, textEditor)) {
+                    vscode.window.showInformationMessage("Done inserting parameter.");
+                }
+                break;
+            case TemplateSectionType.Resources:
+                if (await this.insertResource(template, textEditor)) {
+                    vscode.window.showInformationMessage("Press TAB to move between the tab stops.");
+                }
+                break;
+            case TemplateSectionType.Variables:
+                if (await this.insertVariable(template, textEditor)) {
+                    vscode.window.showInformationMessage("Please type the the value of the variable.");
+                }
+                break;
+            default:
+                vscode.window.showWarningMessage("Unknown insert item type!");
+                return;
+        }
+    }
+
+    private getTemplateObjectPart(template: DeploymentTemplate, templatePart: string): Json.ObjectValue | undefined {
+        let part = this.getTemplatePart(template, templatePart);
+        return Json.asObjectValue(part);
+    }
+
+    private getTemplateArrayPart(template: DeploymentTemplate, templatePart: string): Json.ArrayValue | undefined {
+        let part = this.getTemplatePart(template, templatePart);
+        return Json.asArrayValue(part);
+    }
+
+    private getTemplatePart(template: DeploymentTemplate, templatePart: string): Json.Value | undefined {
+        let rootValue = template.topLevelValue;
+        if (!rootValue) {
+            return undefined;
+        }
+        return rootValue.getPropertyValue(templatePart);
+    }
+
+    private async insertParameter(template: DeploymentTemplate, textEditor: vscode.TextEditor): Promise<boolean> {
+        let name = await this.ui.showInputBox({ prompt: "Name of parameter?" });
+        const parameterType = await this.ui.showQuickPick(getItemType(), { placeHolder: 'Type of parameter?' });
+        let parameter: Parameter = {
+            type: parameterType.value
+        };
+        let defaultValue = await this.ui.showInputBox({ prompt: "Default value? Leave empty for no default value.", });
```

Comment:
suggestion:
parameter.defaultValue  = await this.ui.showInputBox({ prompt: "Default value? Leave empty for no default value.", }) ?? undefined;

Reply:
Your suggestion created an empty defaultValue in the ARM template when that question was left blank. My suggestion did not add it at all when left blank
```
        "noDefaultValue": {
            "type": "string",
            "defaultValue": "",
            "metadata": {
                "description": "Should not have any default value"
            }
        }
```
Because of that I am keeping the original implementation.

Reply:
thanks for catching